File: frontend/src/api/endpoints/deliveries.ts
/** Client de /api/v1/work-orders/.../deliver-with-signature + acuse. */
import api from "@/api/client";

export interface DeliveryRead {
  id: string;
  branch_id: string;
  work_order_id: string;
  customer_name: string;
  customer_id_type: string | null;
  customer_id_number: string | null;
  signature_url: string;
  pdf_url: string | null;
  delivered_at: string;
  delivered_by_id: string | null;
}

export interface DeliverResponse {
  delivery: DeliveryRead;
  work_order: {
    id: string;
    order_number: string;
    status: string;
    branch_id: string;
    closed_at: string | null;
  };
  pdf_url: string | null;
  whatsapp_link: string | null;
}

export interface DeliverPayload {
  customer_name: string;
  customer_id_type?: string | null;
  customer_id_number?: string | null;
  signature_base64: string;
  evidence_ids?: string[];
  notes?: string | null;
  exit_mileage?: number | null;
}

export const deliveriesApi = {
  deliverWithSignature: (workOrderId: string, payload: DeliverPayload) =>
    api
      .post<DeliverResponse>(`/v1/work-orders/${workOrderId}/deliver-with-signature`, payload)
      .then((r) => r.data),

  acusePdfUrl: (workOrderId: string) => `/v1/work-orders/${workOrderId}/acuse.pdf`,
};

export const workshopQAApi = {
  qaPass: (workOrderId: string, note?: string) =>
    api
      .post<{ id: string; status: string; qa: string }>(`/workshop/work-orders/${workOrderId}/qa-pass`, { note })
      .then((r) => r.data),

  qaFail: (workOrderId: string, reason: string) =>
    api
      .post<{ id: string; status: string; qa: string; reason: string }>(
        `/workshop/work-orders/${workOrderId}/qa-fail`,
        { reason },
      )
      .then((r) => r.data),
};

File: app/schemas/deliveries.py
"""Schemas para entregas con firma digital."""
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class DeliverWithSignatureRequest(BaseModel):
    customer_name: str = Field(..., min_length=1, max_length=255)
    customer_id_type: Optional[str] = Field(None, max_length=32)
    customer_id_number: Optional[str] = Field(None, max_length=64)
    signature_base64: str = Field(..., min_length=10)
    evidence_ids: Optional[list[str]] = None
    notes: Optional[str] = None
    exit_mileage: Optional[int] = None


class DeliveryRead(BaseModel):
    id: str
    branch_id: str
    work_order_id: str
    customer_name: str
    customer_id_type: Optional[str] = None
    customer_id_number: Optional[str] = None
    signature_url: str
    pdf_url: Optional[str] = None
    delivered_at: datetime
    delivered_by_id: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class DeliverResponse(BaseModel):
    delivery: DeliveryRead
    work_order: dict
    pdf_url: Optional[str] = None
    whatsapp_link: Optional[str] = None
